policies/module-types/template/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+pub struct TemplateParameters {
+    /// Output file path
+    path: PathBuf,
+    /// Source template path
+    #[serde(skip_serializing_if = "Option::is_none")]
+    template_path: Option<PathBuf>,
+    /// Inlined source template
+    #[serde(skip_serializing_if = "Option::is_none")]
+    template_src: Option<String>,
+    /// Templating engine
+    #[serde(default)]
+    engine: Engine,
+    /// Data to use for templating
+    #[serde(default)]
+    data: Value,
+}
+
+// Module
+
+struct Template {}
+
+impl ModuleType0 for Template {
+    fn metadata(&self) -> ModuleTypeMetadata {
+        let meta = include_str!("../rudder_module_type.yml");
+        let docs = include_str!("../README.md");
+        ModuleTypeMetadata::from_metadata(meta)
+            .expect("invalid metadata")
+            .documentation(docs)
+    }
+
+    fn validate(&self, parameters: &Parameters) -> ValidateResult {
+        // Parse as parameters type
+        let parameters: TemplateParameters =
+            serde_json::from_value(Value::Object(parameters.data.clone()))?;
+
+        match (
+            parameters.template_path.is_some(),
+            parameters.template_src.is_some(),
+        ) {
+            (true, true) => bail!("Only one of 'template_path' and 'template_src' can be provided"),
+            (false, false) => {
+                bail!("Need one of 'template_path' and 'template_src'")
+            }
+            _ => (),
+        }
+
+        Ok(())
+    }
+
+    fn check_apply(&mut self, mode: PolicyMode, parameters: &Parameters) -> CheckApplyResult {
+        assert!(self.validate(parameters).is_ok());
+        let p: TemplateParameters = serde_json::from_value(Value::Object(parameters.data.clone()))?;
+        let output_file = &p.path;
+        let output_file_d = output_file.display();
+
+        // Compute output
+        let output = p
+            .engine
+            .render(p.template_path.as_deref(), p.template_src, p.data)?;
+
+        let already_present = output_file.exists();
+
+        // Check if already correct
+        let mut content = String::new();
+        let already_correct = if already_present {
+            content = read_to_string(output_file)
+                .with_context(|| format!("Failed to read file {output_file_d}"))?;
+            if content == output {
+                true
+            } else {
+                rudder_debug!(
+                    "Output file {} exists but has outdated content",
+                    output_file_d
+                );
+                false
+            }
+        } else {
+            rudder_debug!("Output file {output_file_d} does not exist");
+            false
+        };
+
+        let outcome = match (already_correct, mode) {
+            (true, _) => Outcome::success(),
+            (false, PolicyMode::Audit) => {
+                if already_present {
+                    bail!("Output file {output_file_d} is present but content is not up to date")
```

Comment:
we want a diff here too. We should probably extract the diff itself in a small function.